shachain/producer.go

```diff
@@ -0,0 +1,72 @@
+package shachain
+
+import (
+	"github.com/roasbeef/btcd/wire"
+)
+
+// SecretProducer is an interface which serves as an abstraction over data
+// structure responsible for effecient generating the secrets by given index.
+// The generation of secrets should be made in such way that secret store
+// might efficiently store and retrieve the secrets.
+type SecretProducer interface {
+	// Produce secret by given index.
+	Produce(uint64) (*wire.ShaHash, error)
```

Comment:
I prefer the prior naming schema left over from the `elkrem` package. So rather than `Produce` I think `AtIndex` is a more meaningful name.